test/modules/TestResultCoordinator/Reports/EdgeHubRestartTest/EdgeHubRestartDirectMethodReportGenerator.cs

```diff
@@ -0,0 +1,258 @@
+
+            return new EdgeHubRestartDirectMethodReport(
+                this.TrackingId,
+                this.TestReportType.ToString(),
+                passedDirectMethodCount,
+                this.SenderSource,
+                this.ReceiverSource,
+                senderResultCount,
+                receiverResultCount,
+                edgeHubRestartStatistics.MedianPeriod);
+        }
+
+        async Task<(ulong resultCount, bool hasValue, long sequenceNum)> IterateResultToSequenceNumberAsync(
+            bool hasValue,
+            MoveNextResultAsync MoveNextResultAsync,
+            long targetSequenceNumber,
+            ulong resultCount)
+        {
+            long seqNum = 0;
+
+            while (hasValue && (seqNum < targetSequenceNumber))
+            {
+                (resultCount, hasValue, seqNum) = await MoveNextResultAsync(resultCount);
+            }
+
+            return (resultCount: resultCount,
+                hasValue: hasValue,
+                sequenceNum: seqNum);
+        }
+
+        void ValidateResult(
+            TestOperationResult result,
+            string expectedSource,
+            string testOperationResultType)
+        {
+            if (!result.Source.Equals(expectedSource, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new InvalidDataException($"Result source is '{result.Source}' but expected should be '{expectedSource}'.");
+            }
+
+            if (!result.Type.Equals(testOperationResultType, StringComparison.Ordinal))
+            {
+                throw new InvalidDataException($"Result type is '{result.Type}' but expected should be '{testOperationResultType}'.");
+            }
+        }
+
+        void AddEntryToCompletedStatusHistogram(EdgeHubRestartDirectMethodResult senderResult)
+        {
+            HttpStatusCode completedStatus = senderResult.DirectMethodCompletedStatusCode;
+            TimeSpan completedPeriod = senderResult.DirectMethodCompletedTime - senderResult.EdgeHubRestartedTime;
+            // Try to allocate the list if it is the first time HttpStatusCode shows up
+            this.completedStatusHistogram.TryAdd(completedStatus, new List<TimeSpan>());
+            this.completedStatusHistogram[completedStatus].Add(completedPeriod);
+        }
+
+        long ConvertStringToLong(string result)
+        {
+            long seqNum;
+            long.TryParse(result, out seqNum);
+            return seqNum;
+        }
+
+        private async Task<(ulong resultCount, bool hasValue, long sequenceNumber)> MoveNextSenderResultAsync(ulong senderResultCount)
+        {
+            bool hasValue = await this.SenderTestResults.MoveNextAsync();
+            long seqNum = 0;
+
+            if (!hasValue)
+            {
+                return (resultCount: senderResultCount,
+                    hasValue: hasValue,
+                    sequenceNumber: seqNum);
```

Comment:
should return original sequence number rather than zero.

Reply:
Quote from Bear's reply:

If I want to return a previous value, I'd need to deserialize the result which can be expensive operation given the routine
	if I want to return null, then I need to make sure the caller can handle Null
	if I do Option, then the caller needs to handle option type properly

Why is 0 not a sufficient given the flag hasValue is also a return param?

I mean the seqNum is essentially an Option<> with respect to hasValue flag.

Reply:
That's fair.